File: leetcode/82-remove-duplicates-from-sorted-list-ii.cpp
/*
Given a sorted linked list, delete all nodes that have duplicate numbers, leaving only distinct numbers from the original list.

For example,
Given 1->2->3->3->4->4->5, return 1->2->5.
Given 1->1->1->2->3, return 2->3.
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return head;
        }
        ListNode dummy(0);
        int current_count = 1;
        auto current = head;
        auto to_check = head->next;
        auto tail = &dummy;

        while (to_check != nullptr) {
            if (current->val == to_check->val) {
                ++current_count;
            } else if (current->val < to_check->val) {
                if (current_count == 1) {
                    tail->next = current;
                    tail = tail->next;
                }
                current = to_check;
                current_count = 1;
            }
            to_check = to_check->next;
        }
        if (current_count == 1) {
            tail->next = current;
            tail = tail->next;
        }
        tail->next = nullptr;

        return dummy.next;
    }
};

File: leetcode/82-remove-duplicates-from-sorted-list-ii.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {ListNode}
def delete_duplicates(head)
  return nil if head.nil?
  dummy = ListNode.new 0  
  p = head
  dup = false
  current = dummy
  while (not p.nil?) and (not p.next.nil?)
    while (not p.next.nil?) and p.next.val == p.val
      dup = true
      p = p.next
    end
    unless dup
      current.next = p
      current = current.next
    end
    dup = false
    p = p.next
  end
  current.next = p

  dummy.next
end
